File: src/automation/scripts/shared/timestampUtils.js
/**
 * Unified timestamp utilities for consistent date/time formatting across the system.
 * 
 * This module provides a single source of truth for timestamp generation to fix
 * inconsistencies between webhook, CSV processing, and HCP sync timestamps.
 * 
 * Standard format: ISO 8601 with seconds precision and proper timezone
 * Example: 2025-06-11T23:42:05-07:00
 */

/**
 * Get current timestamp in Arizona timezone with consistent formatting.
 * 
 * Note: This implementation correctly uses timezone-aware dates instead of
 * manually adjusting UTC time.
 * 
 * @returns {string} ISO 8601 formatted timestamp with proper timezone
 *                   Example: "2025-06-11T23:42:05-07:00"
 */
function getArizonaTimestamp() {
  // Create a date in Arizona timezone (MST, UTC-7, no DST)
  const now = new Date();
  
  // Format with Intl.DateTimeFormat for proper timezone handling
  const options = {
    timeZone: 'America/Phoenix',
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
    hour12: false
  };
  
  const formatter = new Intl.DateTimeFormat('en-US', options);
  const parts = formatter.formatToParts(now);
  
  // Build ISO format from parts
  const dateParts = {};
  parts.forEach(part => {
    dateParts[part.type] = part.value;
  });
  
  // Construct ISO string with Arizona timezone offset (-07:00)
  const isoString = `${dateParts.year}-${dateParts.month}-${dateParts.day}T${dateParts.hour}:${dateParts.minute}:${dateParts.second}-07:00`;
  
  return isoString;
}

/**
 * Get current timestamp with space separator (for backward compatibility).
 * 
 * @returns {string} Timestamp with space separator
 *                   Example: "2025-06-11 23:42:05-07:00"
 */
function getArizonaTimestampSpace() {
  return getArizonaTimestamp().replace('T', ' ');
}

/**
 * DEPRECATED: Old function that incorrectly shows UTC timezone.
 * Kept for reference - DO NOT USE in new code.
 * 
 * @deprecated Use getArizonaTimestamp() instead
 */
function getArizonaTimeOld() {
  const now = new Date();
  // This is WRONG - it adjusts the time but shows 'Z' (UTC) suffix
  const arizonaTime = new Date(now.getTime() - (7 * 60 * 60 * 1000));
  return arizonaTime.toISOString();
}

/**
 * Parse an ISO timestamp string and return a Date object
 * 
 * @param {string} timestamp - ISO timestamp string
 * @returns {Date} Date object
 */
function parseTimestamp(timestamp) {
  return new Date(timestamp);
}

/**
 * Format a Date object to Arizona timezone ISO string
 * 
 * @param {Date} date - Date object to format
 * @returns {string} ISO 8601 formatted timestamp
 */
function formatDateToArizona(date) {
  const options = {
    timeZone: 'America/Phoenix',
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
    hour12: false
  };
  
  const formatter = new Intl.DateTimeFormat('en-US', options);
  const parts = formatter.formatToParts(date);
  
  const dateParts = {};
  parts.forEach(part => {
    dateParts[part.type] = part.value;
  });
  
  return `${dateParts.year}-${dateParts.month}-${dateParts.day}T${dateParts.hour}:${dateParts.minute}:${dateParts.second}-07:00`;
}

// Export functions for use in other modules
module.exports = {
  getArizonaTimestamp,
  getArizonaTimestampSpace,
  parseTimestamp,
  formatDateToArizona,
  // Deprecated - do not use
  getArizonaTimeOld
};

// Test the functions if run directly
if (require.main === module) {
  console.log('Timestamp Utils Test:');
  console.log(`Standard format: ${getArizonaTimestamp()}`);
  console.log(`Space format:    ${getArizonaTimestampSpace()}`);
  console.log(`Old format (WRONG): ${getArizonaTimeOld()}`);
  
  console.log('\nParsing test:');
  const testTimestamp = '2025-06-11T23:42:05-07:00';
  const parsed = parseTimestamp(testTimestamp);
  console.log(`  ${testTimestamp} -> ${parsed}`);
  console.log(`  Reformatted: ${formatDateToArizona(parsed)}`);
}

File: .git-rewrite/t/src/automation/scripts/shared/timestamp_utils.py
#!/usr/bin/env python3
"""
Unified timestamp utilities for consistent date/time formatting across the system.

This module provides a single source of truth for timestamp generation to fix
inconsistencies between webhook, CSV processing, and HCP sync timestamps.

Standard format: ISO 8601 with seconds precision and proper timezone
Example: 2025-06-11T23:42:05-07:00
"""

from datetime import datetime, timezone
import pytz

# Arizona timezone (MST, no daylight saving)
ARIZONA_TZ = pytz.timezone('America/Phoenix')

def get_arizona_timestamp():
    """
    Get current timestamp in Arizona timezone with consistent formatting.
    
    Returns:
        str: ISO 8601 formatted timestamp with seconds precision
             Example: "2025-06-11T23:42:05-07:00"
    """
    return datetime.now(ARIZONA_TZ).isoformat(sep='T', timespec='seconds')

def get_arizona_timestamp_space():
    """
    Get current timestamp with space separator (for backward compatibility).
    
    Returns:
        str: Timestamp with space separator and seconds precision
             Example: "2025-06-11 23:42:05-07:00"
    """
    return datetime.now(ARIZONA_TZ).isoformat(sep=' ', timespec='seconds')

def get_arizona_datetime():
    """
    Get current datetime object in Arizona timezone.
    
    Returns:
        datetime: Timezone-aware datetime object
    """
    return datetime.now(ARIZONA_TZ)

def format_timestamp_for_airtable(dt=None):
    """
    Format a datetime for Airtable consistency.
    
    Args:
        dt: datetime object (if None, uses current time)
        
    Returns:
        str: Formatted timestamp for Airtable
    """
    if dt is None:
        dt = get_arizona_datetime()
    elif not dt.tzinfo:
        # If naive datetime, assume it's in Arizona timezone
        dt = ARIZONA_TZ.localize(dt)
    elif dt.tzinfo != ARIZONA_TZ:
        # Convert to Arizona timezone if different
        dt = dt.astimezone(ARIZONA_TZ)
    
    # Use consistent format: ISO with 'T' separator and seconds precision
    return dt.isoformat(sep='T', timespec='seconds')

def parse_airtable_timestamp(timestamp_str):
    """
    Parse a timestamp from Airtable into a datetime object.
    
    Args:
        timestamp_str: String timestamp from Airtable
        
    Returns:
        datetime: Timezone-aware datetime object
    """
    # Handle various formats we might encounter
    if not timestamp_str:
        return None
    
    # Try parsing with dateutil for flexibility
    from dateutil.parser import parse
    dt = parse(timestamp_str)
    
    # Ensure timezone awareness
    if not dt.tzinfo:
        dt = ARIZONA_TZ.localize(dt)
    
    return dt

# JavaScript-compatible function for HCP sync
def get_arizona_timestamp_js():
    """
    Get timestamp in a format compatible with JavaScript Date.toISOString().
    Note: This returns proper Arizona time with correct timezone offset,
    not UTC with manual offset like the current JS implementation.
    
    Returns:
        str: ISO 8601 formatted timestamp
             Example: "2025-06-11T23:42:05-07:00"
    """
    return get_arizona_timestamp()

if __name__ == "__main__":
    # Test the functions
    print("Timestamp Utils Test:")
    print(f"Standard format: {get_arizona_timestamp()}")
    print(f"Space format:    {get_arizona_timestamp_space()}")
    print(f"JS compatible:   {get_arizona_timestamp_js()}")
    print(f"DateTime object: {get_arizona_datetime()}")
    
    # Test parsing
    test_timestamps = [
        "2025-06-11T23:42:05-07:00",
        "2025-06-11 23:42:05-07:00",
        "2025-06-11T23:42:05.123456-07:00",
        "2025-06-11T23:42:05Z"
    ]
    
    print("\nParsing tests:")
    for ts in test_timestamps:
        parsed = parse_airtable_timestamp(ts)
        print(f"  {ts} -> {parsed}")